core/src/main/kotlin/com/strumenta/kolasu/model/Indexing.kt

```diff
@@ -0,0 +1,19 @@
+package com.strumenta.kolasu.model
+
+import java.util.IdentityHashMap
+import java.util.UUID
+
+/**
+ * @param walker a function that generates a sequence of nodes. By default this is the depth-first "walk" method.
+ * @param idProvider a function that takes a node and returns a String identifier representing it
+ * For post-order traversal, take "walkLeavesFirst".
+ * @return walks the whole AST starting from the given node and associates each visited node with a generated id
+ **/
+fun Node.computeIds(
+    walker: (Node) -> Sequence<Node> = Node::walk,
+    idProvider: (Node) -> String = { UUID.randomUUID().toString() }
```

Comment:
For the default implementation I would consider a mechanism that provides sequential IDs, so that they are shorter. The advantage of shorter IDs is that the resulting files are smaller and they are more "human friendly"

Reply:
Do you mean using something similar to a counter?

Reply:
Nice touch to have an idProvider!

Reply:
Yes, a counter would be perfect

Reply:
I have defined an `IdProvider` interface with a `SequentialIdProvider` default implementation for such purpose. Now sequential integers are used in place of random UUIDs. Custom IdProviders can be implemented as lambda, if no local state is required, like so:
```
node.computeIds {
    node -> some_function_returning_string(node) // this will be the node id
}
```
Otherwise, stateful idProviders can be implemented as the default `SequentialIdProvider` or `OnlyReferencedIdProvider`, like:
```
class SequentialIdProvider(private var counter: Int = 0) : IdProvider {
    override fun getId(node: Node): String? {
        return "${this.counter++}"
    }    
}
```

`OnlyReferencedIdProvider` is used to limit id computation to nodes that have at least one ReferenceByName targeting them.